sdks/java/core/src/test/java/org/apache/beam/sdk/util/ApiSurfaceVerificationTest.java

```diff
@@ -0,0 +1,244 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import static org.hamcrest.Matchers.anyOf;
+import static org.junit.Assert.fail;
+
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Predicate;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Ordering;
+import com.google.common.collect.Sets;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nonnull;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.StringDescription;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This test aims at keeping the public API is conformant to a hard-coded policy by
+ * testing whether a package (determined by the location of the test) exposes only white listed
+ * packages/classes.
+ * <p>
+ * Tests that derive from {@link ApiSurfaceVerificationTest} should be placed under the package to
+ * be * tested and implement {@link ApiSurfaceVerificationTest#allowedClasses()}.
+ * Further customization can be done by overriding {@link ApiSurfaceVerificationTest#apiSurface()}.
+ * </p>
+ */
+// based on previous code by @kennknowles and others.
+@RunWith(JUnit4.class)
+public abstract class ApiSurfaceVerificationTest {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ApiSurfaceVerificationTest.class);
+
+  private static class ClassInPackage extends TypeSafeDiagnosingMatcher<Class<?>> {
+
+    private final String packageName;
+
+    private ClassInPackage(final String packageName) {
+      this.packageName = packageName;
+    }
+
+    @Override
+    public void describeTo(final Description description) {
+      description.appendText("Class in package \"");
+      description.appendText(packageName);
+      description.appendText("\"");
+    }
+
+    @Override
+    protected boolean matchesSafely(final Class<?> clazz, final Description mismatchDescription) {
+      return clazz.getName().startsWith(packageName + ".");
+    }
+  }
+
+  protected static Matcher<Class<?>> inPackage(final String packageName) {
+    return new ClassInPackage(packageName);
+  }
+
+  protected abstract Set<Matcher<Class<?>>> allowedClasses();
+
+  private ApiSurface prune(final ApiSurface apiSurface, final Set<String> prunePatterns) {
```

Comment:
This would most naturally be a method of `ApiSurface` or a utilities in a sibling class `ApiSurfaces`.